File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
from collections import defaultdict

def solution(record):
    # 키: 유저 아이디
    # 밸류: 유저 닉네임
    # 다 해놓고 다시 record 뒤져서 행동 출력하기?
    
    answer = []
    nick_dict = defaultdict(str)
    
    for r in record:
        r = r.split()
        
        if r[0] == "Leave":
            continue
        
        nick_dict[r[1]] = r[2]
    
    for r in record:
        r = r.split()
        
        if r[0] == "Enter":
            answer.append(nick_dict[r[1]] + "님이 들어왔습니다.")
        elif r[0] == "Leave":
            answer.append(nick_dict[r[1]] + "님이 나갔습니다.")
    
    return answer

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.js
function solution(record) {
    const answer = [];
    const user_nick = {};
    
    for (let i = 0; i < record.length; i++) {
        const parts = record[i].split(" ");
        const active = parts[0];
        const uid = parts[1];
        
        if (active === "Enter" || active === "Change") {
            user_nick[uid] = parts[2];
        }
    }
    
    for (let i = 0; i < record.length; i++) {
        const parts = record[i].split(" ");
        const active = parts[0];
        const uid = parts[1];
        
        if (active === "Enter") {
            answer.push(`${user_nick[uid]}님이 들어왔습니다.`)
        } else if (active === "Leave") {
            answer.push(`${user_nick[uid]}님이 나갔습니다.`)
        }
    }
    
    return answer;
}